tokio/src/sync/semaphore_ll.rs

```diff
@@ -501,86 +596,149 @@ impl Permit {
-            Pending => {
-                self.state = PermitState::Waiting;
                 Pending
             }
+            Acquired(acquired) => {
+                if acquired >= num_permits {
+                    Ready(Ok(()))
+                } else {
+                    match semaphore.poll_acquire(cx, num_permits - acquired, self)? {
+                        Ready(()) => {
+                            self.state = Acquired(num_permits);
+                            Ready(Ok(()))
+                        }
+                        Pending => {
+                            self.state = Waiting(num_permits);
+                            Pending
+                        }
+                    }
+                }
+            }
         }
     }
 
     /// Try to acquire the permit.
-    pub(crate) fn try_acquire(&mut self, semaphore: &Semaphore) -> Result<(), TryAcquireError> {
+    pub(crate) fn try_acquire(
+        &mut self,
+        num_permits: u16,
+        semaphore: &Semaphore,
+    ) -> Result<(), TryAcquireError> {
+        use PermitState::*;
+
         match self.state {
-            PermitState::Idle => {}
-            PermitState::Waiting => {
+            Waiting(requested) => {
+                // There must be a waiter
                 let waiter = self.waiter.as_ref().unwrap();
 
-                if waiter.acquire2().map_err(to_try_acquire)? {
-                    self.state = PermitState::Acquired;
-                    return Ok(());
+                if requested > num_permits {
+                    let delta = requested - num_permits;
+                    let to_release = waiter.try_dec_permits_to_acquire(delta as usize);
+
+                    semaphore.add_permits(to_release);
+                    self.state = Waiting(num_permits);
+                }
+
+                let res = waiter.permits_to_acquire().map_err(to_try_acquire)?;
+
+                if res == 0 {
+                    if requested < num_permits {
+                        // Try to acquire the additional permits
+                        semaphore.try_acquire(num_permits - requested)?;
+                    }
+
+                    self.state = Acquired(num_permits);
+                    Ok(())
                 } else {
-                    return Err(TryAcquireError::no_permits());
+                    Err(TryAcquireError::NoPermits)
                 }
             }
-            PermitState::Acquired => {
-                return Ok(());
-            }
-        }
-
-        match semaphore.poll_permit(None).map_err(to_try_acquire)? {
-            Ready(()) => {
-                self.state = PermitState::Acquired;
-                Ok(())
+            Acquired(acquired) => {
+                if acquired >= num_permits {
+                    Ok(())
+                } else {
+                    semaphore.try_acquire(num_permits - acquired)?;
+                    self.state = Acquired(num_permits);
+                    Ok(())
+                }
```

Comment:
nit: I think this could be simplified to
```rust
if acquired < num_permits {
    semaphore.try_acquire(num_permits - acquired)?;
    self.state = Acquired(num_permits);
}
Ok(())
```